fix(api): skip picture validation when unchanged

steven-the-qa opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitHub Codespaces.

<!-- Feel free to add any additional description of changes below this line -->

## Problem

<img width="688" height="505" alt="Screenshot 2025-12-23 at 9 13 35 AM" src="https://github.com/user-attachments/assets/b6875bb6-e8d0-401d-bd5c-5e507d6bd8a0" />

<img width="1517" height="979" alt="Screenshot 2025-12-23 at 9 14 01 AM" src="https://github.com/user-attachments/assets/68a4d439-2562-4df6-b825-6be498600809" />


Users would experience a 400 error when updating their profile's "About" field or location without changing their profile picture. This occurred because:

1. The form always sends all fields (`name`, `location`, `picture`, `about`) even when only one field is changed
2. The server was validating the `picture` URL on every update, even when it hadn't changed
3. Some users have profile pictures from services like GitHub (`https://avatars0.githubusercontent.com/u/34585031?v=4`) that don't include file extensions (`.jpg`, `.png`, etc.)
4. The validation requires file extensions, causing legitimate updates to fail with a 400 error

This particularly affected users who:
- Hadn't logged in for a long time and had older avatar URLs
- Only wanted to update their bio or location
- Had GitHub or other avatar URLs without explicit file extensions

## Solution

Modified the `/update-my-about` endpoint to only validate the picture URL when it's actually being changed:

1. **Conditional validation**: The picture is now validated only if the new value differs from the current value in the database
2. **Skip validation for unchanged pictures**: If the picture URL is unchanged, validation is skipped entirely, allowing users to update other fields without errors
3. **Maintain security**: Picture validation still occurs when users actually change their picture, ensuring invalid URLs are still rejected

## Changes Made

- **`api/src/routes/protected/settings.ts`**: Updated the `/update-my-about` handler to compare the incoming picture URL with the current value before validating
- **`api/src/routes/protected/settings.test.ts`**: Added comprehensive API tests covering:
  - Updating location/about when picture is unchanged (including URLs without extensions)
  - Updating only location when picture is unchanged
  - Updating only about when picture is unchanged
  - Verifying that picture validation still works when the picture is actually changed

## Testing

- ✅ Added 4 new API tests covering the fix scenarios
- ✅ Verified that picture validation still works correctly when pictures are changed

## Impact

This fix allows users to update their profile information (About, Location, Name) without encountering validation errors for their existing profile picture, while maintaining security by still validating pictures when they are actually changed.

steven-the-qa committed:
Only validate changed profile data on Save

ojeytonwilliams commented on api/src/routes/protected/settings.ts:
```diff
@@ -523,22 +523,42 @@ ${isLinkSentWithinLimitTTL}`
-        return { message: 'flash.wrong-updating', type: 'danger' } as const;
+      
+      // Only validate picture if it's being changed
+      if (req.body.picture !== undefined) {
```
```suggestion
      if (req.body.picture) {
```
If it's undefined, nothing will be updated (which is fine)

ojeytonwilliams commented on api/src/routes/protected/settings.ts:
```diff
@@ -523,22 +523,42 @@ ${isLinkSentWithinLimitTTL}`
-        void reply.code(400);
-        return { message: 'flash.wrong-updating', type: 'danger' } as const;
+      
+      // Only validate picture if it's being changed
```
```suggestion
      // No need to validate if picture is being deleted.
```

ojeytonwilliams reviewed (changes requested):
Hi @steven-the-qa thanks for this PR and for taking the time to write more tests.

I think we can simplify things a little though, because the client always sends strings, never undefined:

steven-the-qa committed:
Remove redundant unit tests

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

steven-the-qa committed:
Update api/src/routes/protected/settings.ts

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

steven-the-qa committed:
Revert change regarding updateData payload

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

steven-the-qa committed:
Remove redundant var

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

steven-the-qa committed:
improve accuracy of comment

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

steven-the-qa committed:
remove redundant not-undefined condition

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

steven-the-qa committed:
maybe-fix: revert most recent commit to satisfy linter

Update picture validation to check for "not undefined" instead of truthy value.

steven-the-qa committed:
revert reversion

I was wrong -- the issue seems to be Prettier formatting

steven-the-qa committed:
format settings file

steven-the-qa commented on .prettierignore:
```diff
@@ -21,3 +21,5 @@ shared/utils/validate.js
 shared-dist
 dist
+node_modules
+.pnpm-store
```
When I was running Prettier locally to resolve the linter complaint in CI, Cursor recommended adding these to the `.prettierignore`. Any issue with this? Happy to remove.

<img width="620" height="362" alt="Screenshot 2026-01-12 at 2 09 55 PM" src="https://github.com/user-attachments/assets/5a2c31df-6573-41cd-a109-db81aabc6840" />

ShaunSHamilton reviewed (changes requested):
`req.user` in routes behind auth have the full user record.

steven-the-qa committed:
Remove redundant user record search

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

steven-the-qa committed:
remove node_modules and pnpm-store from prettierignore

ShaunSHamilton reviewed (approved):
Thank you for taking the time to contribute! 👍
We are happy to accept these changes and look forward to future contributions. 🎉

ojeytonwilliams reviewed (approved):
LGTM 👍 Thanks!
